File: scripts/resolve-pr-conversation.go
//usr/bin/env go run "$0" "$@"; exit

package main

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"os/exec"
	"strings"
)

// GitHubThread represents a GitHub review thread
type GitHubThread struct {
	ID         string `json:"id"`
	IsResolved bool   `json:"isResolved"`
}

// GitHubError represents a GitHub API error
type GitHubError struct {
	Message string `json:"message"`
}

// GitHubResponse represents the GitHub GraphQL API response
type GitHubResponse struct {
	Data *struct {
		ResolveReviewThread *struct {
			Thread GitHubThread `json:"thread"`
		} `json:"resolveReviewThread,omitempty"`
		AddPullRequestReviewThreadReply *struct {
			Comment struct {
				ID string `json:"id"`
			} `json:"comment"`
		} `json:"addPullRequestReviewThreadReply,omitempty"`
	} `json:"data,omitempty"`
	Errors []GitHubError `json:"errors,omitempty"`
}

// addCommentToThread adds a comment to a specific review thread
func addCommentToThread(threadID, comment string) error {
	fmt.Printf("Adding comment to thread: %s\n", threadID)

	// Escape the comment for GraphQL
	escapedComment := strings.ReplaceAll(comment, `\`, `\\`)
	escapedComment = strings.ReplaceAll(escapedComment, `"`, `\"`)
	escapedComment = strings.ReplaceAll(escapedComment, "\n", `\n`)

	query := fmt.Sprintf(`mutation {
		addPullRequestReviewThreadReply(input: {
			pullRequestReviewThreadId: "%s",
			body: "%s"
		}) {
			comment {
				id
			}
		}
	}`, threadID, escapedComment)

	cmd := exec.Command("gh", "api", "graphql", "-f", "query="+query)
	output, err := cmd.Output()
	if err != nil {
		return fmt.Errorf("failed to execute GraphQL query: %w", err)
	}

	var response GitHubResponse
	if err := json.Unmarshal(output, &response); err != nil {
		return fmt.Errorf("failed to parse response: %w", err)
	}

	if response.Data != nil && response.Data.AddPullRequestReviewThreadReply != nil {
		fmt.Println("✅ Successfully added comment to thread")
		return nil
	}

	if len(response.Errors) > 0 {
		fmt.Fprintln(os.Stderr, "❌ GitHub API errors:")
		for _, err := range response.Errors {
			fmt.Fprintf(os.Stderr, "  - %s\n", err.Message)
		}
		return fmt.Errorf("GitHub API returned errors")
	}

	return fmt.Errorf("unexpected response format")
}

// resolveConversation resolves a GitHub review thread (conversation) using the provided thread ID
func resolveConversation(conversationID, comment string) error {
	if conversationID == "" {
		fmt.Fprintln(os.Stderr, "Error: conversationId is required")
		fmt.Fprintln(os.Stderr, "Usage: go run resolve-pr-conversation.go <conversationId> [comment]")
		fmt.Fprintln(os.Stderr, "")
		fmt.Fprintln(os.Stderr, "If a comment is provided, it will be added to the conversation thread before resolving.")
		return fmt.Errorf("conversationId is required")
	}

	// Add comment to thread if provided
	if comment != "" {
		if err := addCommentToThread(conversationID, comment); err != nil {
			return fmt.Errorf("failed to add comment: %w", err)
		}
	}

	// Resolve the thread
	query := fmt.Sprintf(`mutation { 
		resolveReviewThread(input: {threadId: "%s"}) { 
			thread { id isResolved } 
		} 
	}`, conversationID)

	fmt.Printf("Resolving conversation: %s\n", conversationID)

	cmd := exec.Command("gh", "api", "graphql", "-f", "query="+query)
	output, err := cmd.Output()
	if err != nil {
		return fmt.Errorf("failed to execute GraphQL query: %w", err)
	}

	var response GitHubResponse
	if err := json.Unmarshal(output, &response); err != nil {
		return fmt.Errorf("failed to parse response: %w", err)
	}

	if response.Data != nil && response.Data.ResolveReviewThread != nil {
		thread := response.Data.ResolveReviewThread.Thread
		fmt.Printf("✅ Successfully resolved conversation %s\n", thread.ID)
		status := "Not resolved"
		if thread.IsResolved {
			status = "Resolved"
		}
		fmt.Printf("Status: %s\n", status)
		return nil
	}

	if len(response.Errors) > 0 {
		fmt.Fprintln(os.Stderr, "❌ GitHub API errors:")
		for _, err := range response.Errors {
			fmt.Fprintf(os.Stderr, "  - %s\n", err.Message)
		}
		return fmt.Errorf("GitHub API returned errors")
	}

	return fmt.Errorf("unexpected response format")
}

func main() {
	// Get conversationId and optional comment from command line arguments
	if len(os.Args) < 2 {
		fmt.Fprintln(os.Stderr, "Usage: go run resolve-pr-conversation.go <conversationId> [comment]")
		os.Exit(1)
	}

	conversationID := os.Args[1]
	var comment string
	if len(os.Args) > 2 {
		comment = os.Args[2]
	}

	// Call the function and handle it properly
	if err := resolveConversation(conversationID, comment); err != nil {
		log.Fatalf("Failed to resolve conversation: %v", err)
	}
}

File: scripts/resolve-pr-conversation.ts
#!/usr/bin/env ts-node

import { execSync } from 'child_process';

type GitHubThread = {
  id: string;
  isResolved: boolean;
};

type GitHubError = {
  message: string;
};

type GitHubResponse = {
  data?: {
    resolveReviewThread?: {
      thread: GitHubThread;
    };
    addPullRequestReviewThreadReply?: {
      comment: {
        id: string;
      };
    };
  };
  errors?: GitHubError[];
};

/**
 * Adds a comment to a specific review thread
 * @param threadId - The thread ID to add comment to
 * @param comment - The comment text to add
 */
function addCommentToThread(threadId: string, comment: string): void {
  try {
    console.log(`Adding comment to thread: ${threadId}`);

    const query = `mutation {
      addPullRequestReviewThreadReply(input: {
        pullRequestReviewThreadId: "${threadId}",
        body: "${comment.replace(/\\/g, '\\\\').replace(/"/g, '\\"').replace(/\n/g, '\\n')}"
      }) {
        comment {
          id
        }
      }
    }`;

    const result = execSync(`gh api graphql -f query='${query}'`, {
      encoding: 'utf8',
      stdio: 'pipe',
    });

    const response = JSON.parse(result) as GitHubResponse;

    if (response.data?.addPullRequestReviewThreadReply?.comment) {
      console.log(`✅ Successfully added comment to thread`);
    } else if (response.errors) {
      console.error('❌ GitHub API errors:');
      response.errors.forEach((error: GitHubError) => {
        console.error(`  - ${error.message}`);
      });
      throw new Error('Failed to add comment to thread');
    }
  } catch (error) {
    console.error('❌ Failed to add comment to thread:', error);
    throw error;
  }
}

/**
 * Resolves a GitHub review thread (conversation) using the provided thread ID
 * @param conversationId - The GitHub review thread ID to resolve
 * @param comment - Optional comment to add before resolving
 */
function resolveConversation(conversationId: string, comment?: string): void {
  if (!conversationId) {
    console.error('Error: conversationId is required');
    console.log(
      'Usage: npm run resolve-conversation <conversationId> [comment]'
    );
    console.log('');
    console.log(
      'If a comment is provided, it will be added to the conversation thread before resolving.'
    );
    process.exit(1);
  }

  try {
    // Add comment to thread if provided
    if (comment) {
      addCommentToThread(conversationId, comment);
    }

    // Resolve the thread
    const query = `mutation { 
      resolveReviewThread(input: {threadId: "${conversationId}"}) { 
        thread { id isResolved } 
      } 
    }`;

    console.log(`Resolving conversation: ${conversationId}`);

    const result = execSync(`gh api graphql -f query='${query}'`, {
      encoding: 'utf8',
      stdio: 'pipe',
    });

    const response = JSON.parse(result) as GitHubResponse;

    if (response.data?.resolveReviewThread?.thread) {
      const thread = response.data.resolveReviewThread.thread;
      console.log(`✅ Successfully resolved conversation ${thread.id}`);
      console.log(`Status: ${thread.isResolved ? 'Resolved' : 'Not resolved'}`);
    } else if (response.errors) {
      console.error('❌ GitHub API errors:');
      response.errors.forEach((error: GitHubError) => {
        console.error(`  - ${error.message}`);
      });
      process.exit(1);
    } else {
      console.error('❌ Unexpected response format');
      process.exit(1);
    }
  } catch (error) {
    console.error('❌ Failed to resolve conversation:', error);
    process.exit(1);
  }
}

// Get conversationId and optional comment from command line arguments
const conversationId = process.argv[2];
const comment = process.argv[3];

// Call the function and handle it properly
try {
  resolveConversation(conversationId, comment);
} catch (error) {
  console.error('Failed to resolve conversation:', error);
  process.exit(1);
}
